File: swissknife/intune.py
from typing import Any, Dict, List

from .graph_client import GraphClient


def list_managed_devices(
    client: GraphClient,
    top: int = 50,
) -> List[Dict[str, Any]]:
    """
    Список управляемых устройств Intune.
    GET /deviceManagement/managedDevices
    Требуются: DeviceManagementManagedDevices.Read.All (или ReadWrite.All).
    """
    params = {"$top": top}
    result = client.get("/deviceManagement/managedDevices", params=params)
    return result.get("value", [])


def get_managed_device(
    client: GraphClient,
    device_id: str,
) -> Dict[str, Any]:
    """
    Информация об одном устройстве.
    """
    path = f"/deviceManagement/managedDevices/{device_id}"
    return client.get(path)


def wipe_device(
    client: GraphClient,
    device_id: str,
    keep_enrollment_data: bool = False,
    keep_user_data: bool = False,
) -> Any:
    """
    Wipe устройства (полный/частичный).
    POST /deviceManagement/managedDevices/{id}/wipe
    """
    path = f"/deviceManagement/managedDevices/{device_id}/wipe"
    body = {
        "keepEnrollmentData": keep_enrollment_data,
        "keepUserData": keep_user_data,
        "macOsUnlockCode": None,
    }
    return client.post(path, json=body)


def retire_device(
    client: GraphClient,
    device_id: str,
) -> Any:
    """
    Retire устройства.
    POST /deviceManagement/managedDevices/{id}/retire
    """
    path = f"/deviceManagement/managedDevices/{device_id}/retire"
    return client.post(path, json={})


def remote_lock_device(
    client: GraphClient,
    device_id: str,
) -> Any:
    """
    Remote lock устройства.
    POST /deviceManagement/managedDevices/{id}/remoteLock
    """
    path = f"/deviceManagement/managedDevices/{device_id}/remoteLock"
    return client.post(path, json={})

File: app/internal/services/intune.go
package services

import (
	"encoding/json"
	"net/url"
	"strconv"

	"swissknife-app/internal/session"
)

type IntuneService struct {
	s *session.Session
}

func NewIntuneService(s *session.Session) *IntuneService { return &IntuneService{s: s} }

func (i *IntuneService) Devices(maxItems int) ([]json.RawMessage, error) {
	c, err := i.s.Client()
	if err != nil {
		return nil, err
	}
	return c.ListAll(i.s.Ctx(), "/deviceManagement/managedDevices", topParams(100), maxItems)
}

func (i *IntuneService) Device(deviceID string) (json.RawMessage, error) {
	c, err := i.s.Client()
	if err != nil {
		return nil, err
	}
	var out json.RawMessage
	err = c.Get(i.s.Ctx(), "/deviceManagement/managedDevices/"+url.PathEscape(deviceID), nil, &out)
	return out, err
}

// Wipe is destructive: typed confirm on the device id.
func (i *IntuneService) Wipe(deviceID string, keepEnrollmentData, keepUserData bool, confirm string) error {
	if err := i.s.GuardDestructive(deviceID, confirm); err != nil {
		return err
	}
	c, err := i.s.Client()
	if err != nil {
		return err
	}
	body := map[string]any{
		"keepEnrollmentData": keepEnrollmentData,
		"keepUserData":       keepUserData,
		"macOsUnlockCode":    nil,
	}
	err = c.Post(i.s.Ctx(), "/deviceManagement/managedDevices/"+url.PathEscape(deviceID)+"/wipe", body, nil)
	i.s.Record("intune.wipe", deviceID,
		"keepEnrollment="+strconv.FormatBool(keepEnrollmentData)+" keepUser="+strconv.FormatBool(keepUserData), err)
	return err
}

func (i *IntuneService) Retire(deviceID, confirm string) error {
	if err := i.s.GuardDestructive(deviceID, confirm); err != nil {
		return err
	}
	c, err := i.s.Client()
	if err != nil {
		return err
	}
	err = c.Post(i.s.Ctx(), "/deviceManagement/managedDevices/"+url.PathEscape(deviceID)+"/retire", map[string]any{}, nil)
	i.s.Record("intune.retire", deviceID, "", err)
	return err
}

func (i *IntuneService) RemoteLock(deviceID, confirm string) error {
	if err := i.s.GuardDestructive(deviceID, confirm); err != nil {
		return err
	}
	c, err := i.s.Client()
	if err != nil {
		return err
	}
	err = c.Post(i.s.Ctx(), "/deviceManagement/managedDevices/"+url.PathEscape(deviceID)+"/remoteLock", map[string]any{}, nil)
	i.s.Record("intune.remoteLock", deviceID, "", err)
	return err
}
